From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Groldi <chniucg@gmail.com>
Date: Mon, 11 Sep 2023 00:49:19 +0200
Subject: [PATCH] Integrated base and plugin system into paper


diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
index 3e82ea07ca4194844c5528446e2c4a46ff4acee5..59b80ff226a104c1cfada8a61a850ffe6d5e5cdf 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
@@ -86,6 +86,10 @@ class PaperPluginInstanceManager {
     }
 
     public void loadPlugin(Plugin provided) {
+        // Pulvinar start
+        net.kissenpvp.core.base.KissenCore.getInstance().load(net.kissenpvp.core.base.PluginState.PRE, provided);
+        // Pulvinar end
+
         PluginMeta configuration = provided.getPluginMeta();
 
         this.plugins.add(provided);
@@ -95,6 +99,10 @@ class PaperPluginInstanceManager {
         }
 
         this.dependencyTree.add(configuration);
+
+        // Pulvinar start
+        net.kissenpvp.core.base.KissenCore.getInstance().load(net.kissenpvp.core.base.PluginState.POST, provided);
+        // Pulvinar end
     }
 
     // InvalidDescriptionException is never used, because the old JavaPluginLoader would wrap the exception.
@@ -175,6 +183,10 @@ class PaperPluginInstanceManager {
             return;
         }
 
+        // Pulvinar start
+        net.kissenpvp.core.base.KissenCore.getInstance().enable(net.kissenpvp.core.base.PluginState.PRE, plugin);
+        // Pulvinar end
+
         if (plugin.getPluginMeta() instanceof PluginDescriptionFile) {
             List<Command> bukkitCommands = PluginCommandYamlParser.parse(plugin);
 
@@ -207,6 +219,9 @@ class PaperPluginInstanceManager {
                 return;
                 // Paper end
             }
+            // Pulvinar start
+            net.kissenpvp.core.base.KissenCore.getInstance().enable(net.kissenpvp.core.base.PluginState.POST, plugin);
+            // Pulvinar end
 
             // Perhaps abort here, rather than continue going, but as it stands,
             // an abort is not possible the way it's currently written
@@ -227,6 +242,10 @@ class PaperPluginInstanceManager {
             return;
         }
 
+        // Pulvinar start
+        net.kissenpvp.core.base.KissenCore.getInstance().disable(net.kissenpvp.core.base.PluginState.PRE, plugin);
+        // Pulvinar end
+
         String pluginName = plugin.getPluginMeta().getDisplayName();
 
         try {
@@ -318,6 +337,9 @@ class PaperPluginInstanceManager {
             this.handlePluginException("Error occurred (in the plugin loader) while removing chunk tickets for " + pluginName + " (Is it up to date?)", ex, plugin); // Paper
         }
 
+        // Pulvinar start
+        net.kissenpvp.core.base.KissenCore.getInstance().disable(net.kissenpvp.core.base.PluginState.POST, plugin);
+        // Pulvinar end
     }
 
     // TODO: Implement event part in future patch (paper patch move up, this patch is lower)
diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java
index 097500a59336db1bbfffcd1aa4cff7a8586e46ec..486ebc9c35d6942d7e7e015cbf7fc81f84e5ec23 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java
@@ -243,4 +243,12 @@ public class PaperPluginManagerImpl implements PluginManager, DependencyContext
     public MutableGraph<String> getInstanceManagerGraph() {
         return instanceManager.getDependencyGraph();
     }
+
+    // Pulvinar start
+    @Override
+    public void registerSetting(@NotNull net.kissenpvp.core.api.config.Option<?, ?> setting, @NotNull Plugin plugin)
+    {
+        org.bukkit.Bukkit.getPulvinar().getImplementation(net.kissenpvp.core.api.config.ConfigurationImplementation.class).registerSetting(plugin, setting);
+    }
+    // Pulvinar end
 }
diff --git a/src/main/java/io/papermc/paper/plugin/storage/ServerPluginProviderStorage.java b/src/main/java/io/papermc/paper/plugin/storage/ServerPluginProviderStorage.java
index cb9b13522a976b82bcb71cef486f11f4172e3e99..71b2e53b8dbaf385fa2cadf654a720c889658ab3 100644
--- a/src/main/java/io/papermc/paper/plugin/storage/ServerPluginProviderStorage.java
+++ b/src/main/java/io/papermc/paper/plugin/storage/ServerPluginProviderStorage.java
@@ -35,6 +35,9 @@ public class ServerPluginProviderStorage extends ConfiguredProviderStorage<JavaP
 
             @Override
             public boolean load(PluginProvider<JavaPlugin> provider, JavaPlugin provided) {
+                // Pulvinar start
+                net.kissenpvp.core.base.KissenCore.getInstance().load(net.kissenpvp.core.base.PluginState.PRE, provided);
+                // Pulvinar end
                 // Add it to the map here, we have to run the actual loading logic later.
                 PaperPluginManagerImpl.getInstance().loadPlugin(provided);
                 return true;
@@ -57,6 +60,9 @@ public class ServerPluginProviderStorage extends ConfiguredProviderStorage<JavaP
         try {
             provided.getLogger().info(String.format("Loading server plugin %s", provided.getPluginMeta().getDisplayName()));
             provided.onLoad();
+            // Pulvinar start
+            net.kissenpvp.core.base.KissenCore.getInstance().load(net.kissenpvp.core.base.PluginState.POST, provided);
+            // Pulvinar end
         } catch (Throwable ex) {
             // Don't mark that provider as ERRORED, as this apparently still needs to run the onEnable logic.
             provided.getSLF4JLogger().error("Error initializing plugin '%s' in folder '%s' (Is it up to date?)".formatted(provider.getFileName(), provider.getParentSource()), ex);
diff --git a/src/main/java/net/kissenpvp/pulvinar/base/PulvinarCore.java b/src/main/java/net/kissenpvp/pulvinar/base/PulvinarCore.java
new file mode 100644
index 0000000000000000000000000000000000000000..82cdf5419e0d3bc295f1e45c4f6d14b31813c36d
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/base/PulvinarCore.java
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2023 KissenPvP
+ *
+ * This program is licensed under the Apache License, Version 2.0.
+ *
+ * This software may be redistributed and/or modified under the terms
+ * of the Apache License as published by the Apache Software Foundation,
+ * either version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the Apache
+ * License, Version 2.0 for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * You should have received a copy of the Apache License, Version 2.0
+ * along with this program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+
+package net.kissenpvp.pulvinar.base;
+
+import net.kissenpvp.core.api.base.Implementation;
+import net.kissenpvp.core.api.base.plugin.KissenPlugin;
+import net.kissenpvp.core.api.networking.client.entitiy.ConsoleClient;
+import net.kissenpvp.core.base.KissenCore;
+import net.kissenpvp.pulvinar.api.base.Pulvinar;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.scheduler.CraftScheduler;
+import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+// Pulvinar
+public class PulvinarCore extends KissenCore implements Pulvinar {
+
+    @Override
+    public @NotNull Optional<KissenPlugin> getPlugin(@NotNull String plugin) {
+        return Optional.ofNullable(Bukkit.getPluginManager().getPlugin(plugin));
+    }
+
+    @Override
+    public @NotNull KissenPlugin[] getPlugins() {
+        return Bukkit.getPluginManager().getPlugins();
+    }
+
+    @Override
+    protected void loadImplementations(@NotNull Map<Class<? extends Implementation>, Implementation> loader) {
+        loader.put(net.kissenpvp.core.api.config.ConfigurationImplementation.class, new net.kissenpvp.pulvinar.config.PulvinarConfigurationImplementation());
+        super.loadImplementations(loader);
+    }
+
+    @Override
+    protected void setupDatabase(@NotNull net.kissenpvp.core.api.config.ConfigurationImplementation config, @NotNull net.kissenpvp.core.api.database.connection.DatabaseImplementation database) {
+        super.setupDatabase(config, database);
+    }
+
+    @Override
+    public int getPort() {
+        return Bukkit.getPort();
+    }
+
+    @Override
+    public boolean isOnline(@NotNull UUID uuid) {
+        return Bukkit.getPlayer(uuid)!=null;
+    }
+
+    @Override
+    public boolean isOnline(@NotNull String s) {
+        return Bukkit.getPlayer(s)!=null;
+    }
+
+    @Override
+    public @NotNull ConsoleClient getConsole() {
+        return Bukkit.getConsoleSender();
+    }
+
+    @Override
+    public void runTask(@NotNull KissenPlugin kissenPlugin, @NotNull Runnable runnable) {
+        Bukkit.getScheduler().runTask((Plugin) kissenPlugin, runnable);
+    }
+
+    @Override
+    public void runTask(@NotNull Runnable runnable, int delay, @NotNull String name) {
+        ((CraftScheduler) Bukkit.getScheduler()).scheduleInternalTask(runnable, delay, name);
+    }
+}
diff --git a/src/main/java/net/kissenpvp/pulvinar/config/PulvinarConfigurationImplementation.java b/src/main/java/net/kissenpvp/pulvinar/config/PulvinarConfigurationImplementation.java
new file mode 100644
index 0000000000000000000000000000000000000000..e1d956e9902126f17a172beaef5e83daeed110f9
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/config/PulvinarConfigurationImplementation.java
@@ -0,0 +1,72 @@
+package net.kissenpvp.pulvinar.config;
+
+import net.kissenpvp.core.api.base.plugin.KissenPlugin;
+import net.kissenpvp.core.api.config.Option;
+import net.kissenpvp.core.config.KissenConfigurationImplementation;
+import net.kissenpvp.pulvinar.database.PrivateDatabaseDns;
+import net.kissenpvp.pulvinar.database.UsePrivate;
+import org.bukkit.configuration.file.FileConfiguration;
+import org.bukkit.configuration.file.YamlConfiguration;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.function.Function;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+
+public class PulvinarConfigurationImplementation extends KissenConfigurationImplementation {
+
+    public PulvinarConfigurationImplementation()
+    {
+        super();
+        registerInternalSetting(new UsePrivate());
+        registerInternalSetting(new PrivateDatabaseDns());
+    }
+
+    @Override
+    protected @NotNull File getFile() {
+        return new File("config", "kissen.yml");
+    }
+
+    @Override
+    protected @NotNull File getFile(@NotNull KissenPlugin plugin) {
+        return new File(plugin.getDataFolder(), "config.yml");
+    }
+
+    @Override
+    protected void load(@NotNull File file, @NotNull List<Option<?, ?>> options) {
+        FileConfiguration configuration = YamlConfiguration.loadConfiguration(file);
+        options.forEach(option -> {
+            String path = getPath(option);
+            if (configuration.contains(path)) {
+                setValue(option, Objects.requireNonNull(configuration.getObject(path, option.getConvertClass())));
+                return;
+            }
+            setDefault(option);
+        });
+    }
+
+    @Override
+    protected void write(@NotNull File file, @NotNull List<Option<?, ?>> options) throws IOException {
+        FileConfiguration configuration = YamlConfiguration.loadConfiguration(file);
+
+        options.stream().collect(getCollector()).forEach((group, optionList) -> optionList.forEach(option -> {
+            String path = getPath(option);
+            configuration.set(path, getSerialized(option));
+        }));
+
+        configuration.save(file);
+    }
+
+    @Contract(pure = true, value = "-> new")
+    private @NotNull Collector<Option<?, ?>, ?, TreeMap<String, List<Option<?, ?>>>> getCollector() {
+        Function<Option<?, ?>, String> key = option -> option.getGroup().toLowerCase();
+        return Collectors.groupingBy(key, TreeMap::new, Collectors.toCollection(ArrayList::new));
+    }
+}
diff --git a/src/main/java/net/kissenpvp/pulvinar/database/PrivateDatabaseDns.java b/src/main/java/net/kissenpvp/pulvinar/database/PrivateDatabaseDns.java
new file mode 100644
index 0000000000000000000000000000000000000000..0d9f49011d0d71161e381daf67f7d9034c923501
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/database/PrivateDatabaseDns.java
@@ -0,0 +1,26 @@
+package net.kissenpvp.pulvinar.database;
+
+import net.kissenpvp.core.api.config.options.OptionString;
+import org.jetbrains.annotations.NotNull;
+
+public class PrivateDatabaseDns extends OptionString {
+    @Override
+    public @NotNull String getGroup() {
+        return "database";
+    }
+
+    @Override
+    public @NotNull String getDescription() {
+        return "Second private sql.";
+    }
+
+    @Override
+    public @NotNull String getDefault() {
+        return "jdbc:sqlite:privatekissen.db";
+    }
+
+    @Override
+    public int getPriority() {
+        return 11;
+    }
+}
diff --git a/src/main/java/net/kissenpvp/pulvinar/database/UsePrivate.java b/src/main/java/net/kissenpvp/pulvinar/database/UsePrivate.java
new file mode 100644
index 0000000000000000000000000000000000000000..2992f55cb77bc26243d42991822c5ce73b593a10
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/database/UsePrivate.java
@@ -0,0 +1,26 @@
+package net.kissenpvp.pulvinar.database;
+
+import net.kissenpvp.core.api.config.options.OptionBoolean;
+import org.jetbrains.annotations.NotNull;
+
+public class UsePrivate extends OptionBoolean {
+    @Override
+    public @NotNull String getGroup() {
+        return "database";
+    }
+
+    @Override
+    public @NotNull String getDescription() {
+        return "\nWhether the system should utilize a private SQL database.";
+    }
+
+    @Override
+    public @NotNull Boolean getDefault() {
+        return false;
+    }
+
+    @Override
+    public int getPriority() {
+        return 10;
+    }
+}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 687a3d302cf94ba78099edcb900e734efefb8e70..6dd88101438c49e2fcf0572343b3951643c05cf8 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -192,7 +192,7 @@ import org.bukkit.event.server.ServerLoadEvent;
 
 import co.aikar.timings.MinecraftTimings; // Paper
 
-public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTask> implements ServerInfo, CommandSource, AutoCloseable {
+public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTask> implements ServerInfo, CommandSource, AutoCloseable, net.kissenpvp.core.api.base.KissenServer { // Pulvinar - impement kissen server
 
     private static MinecraftServer SERVER; // Paper
     public static final Logger LOGGER = LogUtils.getLogger();
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 957da64ddd0a56f457122a3f590b45e348f22382..47b56c0530c7c107fd8e82d6c36b44ad2a5ae33c 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -193,6 +193,13 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             DedicatedServer.LOGGER.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
         }
 
+        // Pulvinar start - start system
+        net.kissenpvp.pulvinar.base.PulvinarCore pulvinarCore = new net.kissenpvp.pulvinar.base.PulvinarCore();
+        org.bukkit.Bukkit.setPulvinar(pulvinarCore);
+        //system start is asynchronous
+        java.util.concurrent.CompletableFuture<Void> systemStart = java.util.concurrent.CompletableFuture.runAsync(pulvinarCore::initialize);
+        // Pulvinar stop
+
         // Paper start - detect running as root
         if (io.papermc.paper.util.ServerEnvironment.userIsRootOrAdmin()) {
             DedicatedServer.LOGGER.warn("****************************");
@@ -227,6 +234,11 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         if (this.convertOldUsers()) {
             this.getProfileCache().save(false); // Paper
         }
+
+        // Pulvinar start - wait for system start
+        systemStart.join();
+        // Pulvinar end
+
         this.getPlayerList().loadAndSaveFiles(); // Must be after convertNames
         // Paper end - fix converting txt to json file
         org.spigotmc.WatchdogThread.doStart(org.spigotmc.SpigotConfig.timeoutTime, org.spigotmc.SpigotConfig.restartOnCrash); // Paper - start watchdog thread
@@ -342,6 +354,9 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             DedicatedServer.LOGGER.warn("To change this, set \"online-mode\" to \"true\" in the server.properties file.");
         }
 
+        // Pulvinar start
+        pulvinarCore.start();
+        // Pulvinar end
 
         if (!OldUsersConverter.serverReadyAfterUserconversion(this)) {
             return false;
