From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo <chniucg@gmail.com>
Date: Sun, 24 Sep 2023 13:28:51 +0200
Subject: [PATCH] Include scalable language framework


diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java
index 341bd527b6d5a955f23f22b352d14a4df964b353..4b011884c81b3056ffd509d37021aef584692b3f 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginManagerImpl.java
@@ -254,5 +254,10 @@ public class PaperPluginManagerImpl implements PluginManager, DependencyContext
     public void registerPlayerSetting(@NotNull net.kissenpvp.core.api.user.playersettting.AbstractPlayerSetting<?, ? extends org.bukkit.OfflinePlayer> playerSetting, @NotNull Plugin plugin) {
         org.bukkit.Bukkit.getPulvinar().getImplementation(net.kissenpvp.core.api.user.UserImplementation.class).registerPlayerSetting(plugin, playerSetting);
     }
+
+    @Override
+    public void registerTranslation(@NotNull String key, @NotNull java.text.MessageFormat defaultMessage, @NotNull Plugin plugin) {
+        org.bukkit.Bukkit.getPulvinar().getImplementation(net.kissenpvp.core.api.message.localization.LocalizationImplementation.class).register(plugin, key, defaultMessage);
+    }
     // Pulvinar end
 }
diff --git a/src/main/java/net/kissenpvp/pulvinar/base/PulvinarCore.java b/src/main/java/net/kissenpvp/pulvinar/base/PulvinarCore.java
index f7c4d650a5299d2daff37176472a155d963d8b9f..8988db1efa131be485a38492682c311171dcb678 100644
--- a/src/main/java/net/kissenpvp/pulvinar/base/PulvinarCore.java
+++ b/src/main/java/net/kissenpvp/pulvinar/base/PulvinarCore.java
@@ -54,12 +54,6 @@ public class PulvinarCore extends KissenCore implements Pulvinar {
         loader.put(net.kissenpvp.pulvinar.api.permission.PermissionImplementation.class, permissionImplementation);
         loader.put(net.kissenpvp.core.permission.InternalPermissionImplementation.class, permissionImplementation);
         loader.put(net.kissenpvp.core.api.message.localization.LocalizationImplementation.class, new net.kissenpvp.pulvinar.message.localization.PulvinarLocalizationImplementation());
-        net.kissenpvp.pulvinar.command.PulvinarCommandImplementation commandImplementation = new net.kissenpvp.pulvinar.command.PulvinarCommandImplementation();
-        loader.put(net.kissenpvp.pulvinar.api.command.CommandImplementation.class, commandImplementation);
-        loader.put(net.kissenpvp.core.command.InternalCommandImplementation.class, commandImplementation);
-        loader.put(net.kissenpvp.core.command.confirmation.KissenConfirmationImplementation.class, new net.kissenpvp.pulvinar.command.confirmation.PulvinarConfirmationImplementation());
-        loader.put(net.kissenpvp.pulvinar.api.ban.BanImplementation.class, new net.kissenpvp.pulvinar.ban.PulvinarBanImplementation());
-        loader.put(net.kissenpvp.core.event.EventImplementation.class, new net.kissenpvp.pulvinar.event.KissenEventImplementation());
         super.loadImplementations(loader);
     }
 
diff --git a/src/main/java/net/kissenpvp/pulvinar/message/localization/PulvinarLocalizationImplementation.java b/src/main/java/net/kissenpvp/pulvinar/message/localization/PulvinarLocalizationImplementation.java
new file mode 100644
index 0000000000000000000000000000000000000000..af5fbd3c197f257f532cc8fefd7657490c993dc3
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/message/localization/PulvinarLocalizationImplementation.java
@@ -0,0 +1,23 @@
+/*
+ * Copyright (C) 2023 KissenPvP
+ *
+ * This program is licensed under the Apache License, Version 2.0.
+ *
+ * This software may be redistributed and/or modified under the terms
+ * of the Apache License as published by the Apache Software Foundation,
+ * either version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the Apache
+ * License, Version 2.0 for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * You should have received a copy of the Apache License, Version 2.0
+ * along with this program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+
+package net.kissenpvp.pulvinar.message.localization;
+
+import net.kissenpvp.core.message.localization.KissenLocalizationImplementation;
+
+public class PulvinarLocalizationImplementation extends KissenLocalizationImplementation { }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 045a9ebb00da27b0f73b0e4ea40f3f3f626521a1..bbf34d35ba93b5b07eb1c25cba9a4eba19a8b934 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2350,6 +2350,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         if (this.language == null || !this.language.equals(clientOptions.language())) { // Paper
             PlayerLocaleChangeEvent event = new PlayerLocaleChangeEvent(this.getBukkitEntity(), clientOptions.language());
             this.server.server.getPluginManager().callEvent(event);
+            updateLocale(clientOptions.language()); // Pulvinar
         }
         // CraftBukkit end
         // Paper start - don't call options events on login
@@ -3117,4 +3118,11 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         this.ramBar = ramBar;
     }
     // Purpur end
+
+    // Pulvinar start
+    public void updateLocale(@org.jetbrains.annotations.NotNull String locale)
+    {
+        ((net.kissenpvp.core.user.KissenPublicUser<?>) getBukkitEntity().getUser()).updateLocale(locale);
+    }
+    // Pulvinar end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 87140f4e291d996d1e5dac3277e519d7968d4199..4b2361f0646e5ef484c1cd0eae66a3d8d4e6a79d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3372,7 +3372,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public void openBook(final net.kyori.adventure.inventory.Book book) {
-        final java.util.Locale locale = this.getHandle().adventure$locale;
+        final java.util.Locale locale = getCurrentLocale(); //this.getHandle().adventure$locale; // Pulvinar
         final net.minecraft.world.item.ItemStack item = io.papermc.paper.adventure.PaperAdventure.asItemStack(book, locale);
         final ServerPlayer player = this.getHandle();
         final ServerGamePacketListenerImpl connection = player.connection;
