From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Groldi <chniucg@gmail.com>
Date: Sun, 10 Sep 2023 15:52:02 +0200
Subject: [PATCH] Adjusted gradle script


diff --git a/build.gradle.kts b/build.gradle.kts
index 149f9088fe806467656e8b1c4157df60fda69ba7..e46fd215fea2be1a5667f40cef4de401592ce01f 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -51,6 +51,10 @@ dependencies {
     implementation("org.ow2.asm:asm-commons:9.4")
     // Paper end
 
+    // KissenPaper start
+    implementation(project(":kissen-api") )
+    // KissenPaper end
+
     api("org.apache.maven:maven-resolver-provider:3.8.5")
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
     compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
