From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Groldi <chniucg@gmail.com>
Date: Sun, 24 Sep 2023 16:06:01 +0200
Subject: [PATCH] Integrated command system from kissen into paper


diff --git a/src/main/java/net/kissenpvp/paper/api/command/KissenCommand.java b/src/main/java/net/kissenpvp/paper/api/command/KissenCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..0b37b7c0653395d7aa554846cde0e9254a11c93d
--- /dev/null
+++ b/src/main/java/net/kissenpvp/paper/api/command/KissenCommand.java
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2023 KissenPvP
+ *
+ * This program is licensed under the Apache License, Version 2.0.
+ *
+ * This software may be redistributed and/or modified under the terms
+ * of the Apache License as published by the Apache Software Foundation,
+ * either version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the Apache
+ * License, Version 2.0 for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * You should have received a copy of the Apache License, Version 2.0
+ * along with this program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+
+package net.kissenpvp.paper.api.command;
+
+import org.bukkit.command.Command;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Represents an abstract, base command structure for the Kissen framework.
+ * <p>
+ * This base command class holds essential information about a command and serves as
+ * a foundation for all specific commands within the Kissen command hierarchy.
+ * The structure encapsulates the basic behavior and properties of a command,
+ * such as its name.
+ * <p>
+ * Subclasses should extend this class to provide specific implementations for each command.
+ *
+ * @see Command
+ */
+public abstract class KissenCommand extends Command {
+
+    /**
+     * Constructs a new command with the specified name.
+     *
+     * @param name The name of the command. It must not be null.
+     *             This is the primary identifier for the command and is used in command
+     *             processing to map user inputs to the corresponding command.
+     * @throws NullPointerException if the {@code name} parameter is null.
+     * @see Command
+     */
+    protected KissenCommand(@NotNull String name) {
+        super(name);
+    }
+}
diff --git a/src/main/java/net/kissenpvp/paper/api/command/PaperArgumentParser.java b/src/main/java/net/kissenpvp/paper/api/command/PaperArgumentParser.java
new file mode 100644
index 0000000000000000000000000000000000000000..0dc73860f7084c32a4103a2facb8b5b5426d3d25
--- /dev/null
+++ b/src/main/java/net/kissenpvp/paper/api/command/PaperArgumentParser.java
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2023 KissenPvP
+ *
+ * This program is licensed under the Apache License, Version 2.0.
+ *
+ * This software may be redistributed and/or modified under the terms
+ * of the Apache License as published by the Apache Software Foundation,
+ * either version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the Apache
+ * License, Version 2.0 for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * You should have received a copy of the Apache License, Version 2.0
+ * along with this program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+
+package net.kissenpvp.paper.api.command;
+
+import net.kissenpvp.core.api.command.ArgumentParser;
+import org.bukkit.command.CommandSender;
+
+/**
+ * Interface for a Paper-specific argument parser.
+ * <p>
+ * This interface extends the {@link ArgumentParser} and refines the server entity type to
+ * {@link CommandSender}. It's used for parsing command arguments within the Paper API context.
+ * <p>
+ * This interface does not add any additional methods; it merely specializes the
+ * {@link ArgumentParser} for use with {@link CommandSender} in the Paper API context.
+ * <p>
+ * All methods inherited from {@link ArgumentParser} maintain their original functionality
+ * with the specialized type {@link CommandSender}.
+ *
+ * @param <T> the type of object that the parser can parse.
+ * @see ArgumentParser
+ * @see CommandSender
+ */
+public interface PaperArgumentParser<T> extends ArgumentParser<T, CommandSender> { }
diff --git a/src/main/java/net/kissenpvp/paper/api/command/PaperCommandExceptionHandler.java b/src/main/java/net/kissenpvp/paper/api/command/PaperCommandExceptionHandler.java
new file mode 100644
index 0000000000000000000000000000000000000000..a3e32fb7c82aeaaebd8f1ec73f12ddf066367613
--- /dev/null
+++ b/src/main/java/net/kissenpvp/paper/api/command/PaperCommandExceptionHandler.java
@@ -0,0 +1,23 @@
+package net.kissenpvp.paper.api.command;
+
+import net.kissenpvp.core.api.command.exception.CommandExceptionHandler;
+import org.bukkit.command.CommandSender;
+
+/**
+ * Interface for Paper command exception handler.
+ * <p>
+ * This interface extends the {@link CommandExceptionHandler} and narrows down
+ * the {@code ServerEntity} type to {@link CommandSender}. It's used for handling exceptions
+ * specifically occurring during the execution of commands within the Paper API context.
+ * <p>
+ * This interface does not add any additional methods; it merely specializes the
+ * {@link CommandExceptionHandler} for use with {@link CommandSender} in the Paper API context.
+ * <p>
+ * All methods inherited from {@link CommandExceptionHandler} maintain their original functionality
+ * with the specialized type {@link CommandSender}.
+ *
+ * @param <T> the type of exception this handler can handle
+ * @see CommandExceptionHandler
+ * @see CommandSender
+ */
+public interface PaperCommandExceptionHandler<T extends Throwable> extends CommandExceptionHandler<T, CommandSender> { }
diff --git a/src/main/java/net/kissenpvp/paper/api/command/PaperCommandImplementation.java b/src/main/java/net/kissenpvp/paper/api/command/PaperCommandImplementation.java
new file mode 100644
index 0000000000000000000000000000000000000000..a29d1dac0b3a8737af4e2376ac894bb919fa4416
--- /dev/null
+++ b/src/main/java/net/kissenpvp/paper/api/command/PaperCommandImplementation.java
@@ -0,0 +1,42 @@
+/*
+ * Copyright (C) 2023 KissenPvP
+ *
+ * This program is licensed under the Apache License, Version 2.0.
+ *
+ * This software may be redistributed and/or modified under the terms
+ * of the Apache License as published by the Apache Software Foundation,
+ * either version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the Apache
+ * License, Version 2.0 for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * You should have received a copy of the Apache License, Version 2.0
+ * along with this program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+
+package net.kissenpvp.paper.api.command;
+
+import net.kissenpvp.core.api.base.Implementation;
+import net.kissenpvp.core.api.base.plugin.KissenPlugin;
+import net.kissenpvp.core.api.command.CommandHandler;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Optional;
+
+/**
+ * Interface for the specific implementation of command operations for the Paper API.
+ * <p>
+ * This interface extends the {@link Implementation} and provides methods related to the command handling
+ * in the context of the Paper API. It outlines the operations for command retrieval, registration, tab completion, and de-registration.
+ *
+ * @see Implementation
+ */
+public interface PaperCommandImplementation extends Implementation
+{
+
+    @NotNull Optional<CommandHandler<CommandSender, KissenCommand>> getPluginHandler(@NotNull KissenPlugin kissenPlugin);
+
+}
diff --git a/src/main/java/org/bukkit/plugin/PluginManager.java b/src/main/java/org/bukkit/plugin/PluginManager.java
index 31b9002187aab40abeb6627c0b0d1ee60c575e5b..843968695b50a7ab3c384cd3181e906e525968d3 100644
--- a/src/main/java/org/bukkit/plugin/PluginManager.java
+++ b/src/main/java/org/bukkit/plugin/PluginManager.java
@@ -325,5 +325,10 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
 
     void registerTranslation(@NotNull String key, @NotNull java.text.MessageFormat defaultMessage, @NotNull Plugin plugin);
 
+    void registerCommand(@NotNull Plugin plugin, @NotNull Object... objects);
+
+    void registerExceptionHandler(@NotNull net.kissenpvp.core.api.base.ExceptionHandler<?> exceptionHandler, @NotNull Plugin plugin);
+
+    <T> void registerParser(@NotNull Class<T> type, @NotNull net.kissenpvp.paper.api.command.PaperArgumentParser<T> parser, @NotNull Plugin plugin);
     // KissenPaper end
 }
diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index 3a8ca4bf5c538c793072720159393c5644d42185..e5d4f7df94c73cd2f9b700d8dafc91024698d681 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -994,5 +994,20 @@ public final class SimplePluginManager implements PluginManager {
     public void registerTranslation(@NotNull String key, @NotNull java.text.MessageFormat defaultMessage, @NotNull Plugin plugin) {
         org.bukkit.Bukkit.getKissen().getImplementation(net.kissenpvp.core.api.message.localization.LocalizationImplementation.class).register(plugin, key, defaultMessage);
     }
+
+    @Override
+    public void registerCommand(@NotNull Plugin plugin, @NotNull Object... objects) {
+        org.bukkit.Bukkit.getKissen().getImplementation(net.kissenpvp.paper.api.command.PaperCommandImplementation.class).getPluginHandler(plugin).ifPresent(handler -> handler.registerCommand(objects));
+    }
+
+    @Override
+    public void registerExceptionHandler(@NotNull net.kissenpvp.core.api.base.ExceptionHandler<?> exceptionHandler, @NotNull Plugin plugin) {
+        org.bukkit.Bukkit.getKissen().getImplementation(net.kissenpvp.paper.api.command.PaperCommandImplementation.class).getPluginHandler(plugin).ifPresent(handler -> handler.registerExceptionHandler(exceptionHandler));
+    }
+
+    @Override
+    public <T> void registerParser(@NotNull Class<T> type, @NotNull net.kissenpvp.paper.api.command.PaperArgumentParser<T> parser, @NotNull Plugin plugin) {
+        org.bukkit.Bukkit.getKissen().getImplementation(net.kissenpvp.paper.api.command.PaperCommandImplementation.class).getPluginHandler(plugin).ifPresent(handler -> handler.registerParser(type, parser));
+    }
     // KissenPaper start
 }
