From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Groldi <chniucg@gmail.com>
Date: Fri, 22 Sep 2023 17:59:30 +0200
Subject: [PATCH] Enhanced ban system


diff --git a/src/main/java/net/kissenpvp/paper/api/ban/PaperBanImplementation.java b/src/main/java/net/kissenpvp/paper/api/ban/PaperBanImplementation.java
index 42f8ff65bdd766512d8b2b7384fd6523ba001dd9..7203d8b06838f110d14fc2b4aa28efde9b13727c 100644
--- a/src/main/java/net/kissenpvp/paper/api/ban/PaperBanImplementation.java
+++ b/src/main/java/net/kissenpvp/paper/api/ban/PaperBanImplementation.java
@@ -33,13 +33,13 @@ import java.util.UUID;
 
 public interface PaperBanImplementation extends BanImplementation<PaperBan, PaperPunishment>
 {
-    @NotNull @Unmodifiable Set<PaperPunishment> getPlayerBanSet(@NotNull Context context) throws BackendException;
+    @NotNull @Unmodifiable Set<PaperPunishment> getPunishmentSet(@NotNull Context context) throws BackendException;
 
-    @NotNull @Unmodifiable Set<PaperPunishment> getPlayerBanSet(@NotNull UUID totalID, @NotNull Context context) throws BackendException;
+    @NotNull @Unmodifiable Set<PaperPunishment> getPunishmentSet(@NotNull UUID totalID, @NotNull Context context) throws BackendException;
 
-    @NotNull Optional<PaperPunishment> getCurrentBan(@NotNull UUID totalID, @NotNull Context context) throws BackendException;
+    @NotNull Optional<PaperPunishment> getLatestPunishment(@NotNull UUID totalID, @NotNull Context context) throws BackendException;
 
-    @NotNull Optional<PaperPunishment> getCurrentBan(@NotNull UUID totalID, @NotNull BanType banType, @NotNull Context context) throws BackendException;
+    @NotNull Optional<PaperPunishment> getLatestPunishment(@NotNull UUID totalID, @NotNull BanType banType, @NotNull Context context) throws BackendException;
 
     @NotNull PaperPunishment punish(@NotNull UUID totalID, @NotNull PaperBan ban, @NotNull BanOperator banOperator, @Nullable Component reason, @NotNull Context context) throws BackendException;
 }
diff --git a/src/main/java/net/kissenpvp/paper/api/networking/client/entity/PaperOnlinePlayerClient.java b/src/main/java/net/kissenpvp/paper/api/networking/client/entity/PaperOnlinePlayerClient.java
index e7767de973eb0962b41587c10e0ac03abd9e9cea..63e5c90d3df5f280fa278b1091b609928749c44c 100644
--- a/src/main/java/net/kissenpvp/paper/api/networking/client/entity/PaperOnlinePlayerClient.java
+++ b/src/main/java/net/kissenpvp/paper/api/networking/client/entity/PaperOnlinePlayerClient.java
@@ -24,9 +24,12 @@ import net.kissenpvp.core.api.networking.client.entitiy.ServerEntity;
 import net.kissenpvp.paper.api.ban.PaperPunishment;
 import net.kissenpvp.paper.api.permission.PaperPermission;
 import net.kissenpvp.paper.api.user.rank.PaperPlayerRank;
+import net.kyori.adventure.text.Component;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 public interface PaperOnlinePlayerClient extends OnlinePlayerClient<PaperPermission, PaperPlayerRank, PaperPunishment> {
 
     int warn(@NotNull Ban ban, @NotNull ServerEntity warnOperator);
+    int warn(@NotNull Ban ban, @NotNull ServerEntity warnOperator, @Nullable Component reason);
 }
