From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Groldi <chniucg@gmail.com>
Date: Sun, 13 Aug 2023 04:43:23 +0200
Subject: [PATCH] Reworked user system


diff --git a/src/main/java/net/kissenpvp/paper/api/permission/PaperPermissible.java b/src/main/java/net/kissenpvp/paper/api/permission/PaperPermissible.java
index 21037d40a25d1fe8416442dbbb7c2ca42846cece..b06ec685676f83c696a18a22b6856db08f4731ef 100644
--- a/src/main/java/net/kissenpvp/paper/api/permission/PaperPermissible.java
+++ b/src/main/java/net/kissenpvp/paper/api/permission/PaperPermissible.java
@@ -3,4 +3,5 @@ package net.kissenpvp.paper.api.permission;
 import org.bukkit.permissions.Permissible;
 
 public interface PaperPermissible extends Permissible, PaperPermissionEntry {
+
 }
