From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Sun, 7 Jan 2024 15:22:01 +0100
Subject: [PATCH] Integrated event system from kissen into paper


diff --git a/src/main/java/org/bukkit/event/Event.java b/src/main/java/org/bukkit/event/Event.java
index 8ec56cd6b8e0f5c5dd8c7c88b4671e18dcf109d0..e2db4b4e3c04b408a5ff5d5a464b2b7f26d44a46 100644
--- a/src/main/java/org/bukkit/event/Event.java
+++ b/src/main/java/org/bukkit/event/Event.java
@@ -12,7 +12,8 @@ import org.jetbrains.annotations.NotNull;
  * @see PluginManager#callEvent(Event)
  * @see PluginManager#registerEvents(Listener,Plugin)
  */
-public abstract class Event {
+public abstract class Event implements net.kissenpvp.core.api.event.EventClass
+{
     private String name;
     private final boolean async;
 
diff --git a/src/main/java/org/bukkit/plugin/PluginManager.java b/src/main/java/org/bukkit/plugin/PluginManager.java
index f631f2e890136721d847dd959d038c989ffab926..8e772a79ec5d5f74c3e09b325338b5c9b1229adc 100644
--- a/src/main/java/org/bukkit/plugin/PluginManager.java
+++ b/src/main/java/org/bukkit/plugin/PluginManager.java
@@ -112,6 +112,13 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
      */
     public void callEvent(@NotNull net.kissenpvp.core.api.event.EventClass event) throws IllegalStateException; // KissenPaper - integrate unified event system
 
+    // KissenPaper start - keep compatibility with plugins
+    default void registerEvents(@NotNull Listener listener, @NotNull Plugin plugin)
+    {
+        registerEvents((net.kissenpvp.core.api.event.EventListener<?>) listener, plugin);
+    }
+    // KissenPaper end - keep compatibility with plugins
+
     /**
      * Registers all the events in the given listener class
      *
