From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Groldi <chniucg@gmail.com>
Date: Tue, 15 Aug 2023 22:32:24 +0200
Subject: [PATCH] Integrated Velocity into the project


diff --git a/src/main/java/net/kissenpvp/paper/api/ban/PaperBanImplementation.java b/src/main/java/net/kissenpvp/paper/api/ban/PaperBanImplementation.java
index 27539701fb0071e4751e9c47f449145579e2dc5c..fad5fec3b2d19ac49236959ee310e8ccbcbcabad 100644
--- a/src/main/java/net/kissenpvp/paper/api/ban/PaperBanImplementation.java
+++ b/src/main/java/net/kissenpvp/paper/api/ban/PaperBanImplementation.java
@@ -35,9 +35,9 @@ public interface PaperBanImplementation extends BanImplementation<PaperBan, Pape
 
     @NotNull @Unmodifiable Set<PaperPunishment> getPlayerBanSet(@NotNull UUID totalID, @NotNull Context context) throws BackendException;
 
-    @NotNull Optional<PaperPunishment> getCurrentBan(@NotNull UUID totalID, Context context) throws BackendException;
+    @NotNull Optional<PaperPunishment> getCurrentBan(@NotNull UUID totalID, @NotNull Context context) throws BackendException;
 
-    @NotNull Optional<PaperPunishment> getCurrentBan(@NotNull UUID totalID, @NotNull BanType banType, Context context) throws BackendException;
+    @NotNull Optional<PaperPunishment> getCurrentBan(@NotNull UUID totalID, @NotNull BanType banType, @NotNull Context context) throws BackendException;
 
     @NotNull PaperPunishment punish(@NotNull UUID totalID, @NotNull PaperBan ban, @NotNull BanOperator banOperator, @Nullable Component reason, @NotNull Context context) throws BackendException;
 }
diff --git a/src/main/java/net/kissenpvp/paper/api/base/PaperKissen.java b/src/main/java/net/kissenpvp/paper/api/base/PaperKissen.java
index d6f3c5ed05fc0011ce4618915948d5c0d6db0f02..b5772f9830125732fde292a7f8a99ee19cf869f8 100644
--- a/src/main/java/net/kissenpvp/paper/api/base/PaperKissen.java
+++ b/src/main/java/net/kissenpvp/paper/api/base/PaperKissen.java
@@ -18,8 +18,9 @@ package net.kissenpvp.paper.api.base;
 
 import net.kissenpvp.core.api.base.Kissen;
 import net.kissenpvp.core.api.database.meta.ObjectMeta;
+import org.jetbrains.annotations.NotNull;
 
 public interface PaperKissen extends Kissen {
 
-    ObjectMeta getPrivateMeta();
+    @NotNull ObjectMeta getPrivateMeta();
 }
diff --git a/src/main/java/net/kissenpvp/paper/api/data/LocationNode.java b/src/main/java/net/kissenpvp/paper/api/data/LocationNode.java
index 3151ef92a8bfd02b55b49d78f6a0b77b4d441d7a..1b84a440b230e99f3c0b70e2b18d39134a7f6196 100644
--- a/src/main/java/net/kissenpvp/paper/api/data/LocationNode.java
+++ b/src/main/java/net/kissenpvp/paper/api/data/LocationNode.java
@@ -24,7 +24,7 @@ import org.jetbrains.annotations.NotNull;
 import java.util.Objects;
 
 
-public record LocationNode(String world, double x, double y, double z, float pitch, float yaw)
+public record LocationNode(@NotNull String world, double x, double y, double z, float pitch, float yaw)
 {
 
     public LocationNode(@NotNull Location location)
diff --git a/src/main/java/net/kissenpvp/paper/api/message/SystemMessageEvent.java b/src/main/java/net/kissenpvp/paper/api/message/SystemMessageEvent.java
index 0be13821ac561b8f9a2eccc8260c2464993be085..a21f56229b32c5bec88fbb4761883f6f9ea4d444 100644
--- a/src/main/java/net/kissenpvp/paper/api/message/SystemMessageEvent.java
+++ b/src/main/java/net/kissenpvp/paper/api/message/SystemMessageEvent.java
@@ -41,15 +41,15 @@ public class SystemMessageEvent extends Event implements Cancellable {
         this.cancelled = cancel;
     }
 
-    public Player getReceiver() {
+    public @NotNull Player getReceiver() {
         return receiver;
     }
 
-    public void setComponent(Component component) {
+    public void setComponent(@NotNull Component component) {
         this.component = component;
     }
 
-    public Component getComponent() {
+    public @NotNull Component getComponent() {
         return component;
     }
 }
diff --git a/src/main/java/net/kissenpvp/paper/api/networking/client/entity/PaperPlayerClient.java b/src/main/java/net/kissenpvp/paper/api/networking/client/entity/PaperPlayerClient.java
index c3fc78944bc8e28791d4a2c2844474d245d533e0..033c1840465eb9389ad714c78da691c942d06309 100644
--- a/src/main/java/net/kissenpvp/paper/api/networking/client/entity/PaperPlayerClient.java
+++ b/src/main/java/net/kissenpvp/paper/api/networking/client/entity/PaperPlayerClient.java
@@ -16,10 +16,7 @@
 
 package net.kissenpvp.paper.api.networking.client.entity;
 
-import net.kissenpvp.core.api.ban.Ban;
-import net.kissenpvp.core.api.ban.Punishment;
 import net.kissenpvp.core.api.networking.client.entitiy.PlayerClient;
-import net.kissenpvp.core.api.networking.client.entitiy.ServerEntity;
 import net.kissenpvp.core.api.user.User;
 import net.kissenpvp.core.api.user.suffix.Suffix;
 import net.kissenpvp.paper.api.ban.PaperPunishment;
@@ -30,9 +27,7 @@ import net.kissenpvp.paper.api.permission.PaperPermission;
 import net.kissenpvp.paper.api.permission.PermissibleOverriddenException;
 import net.kissenpvp.paper.api.user.rank.PaperPlayerRank;
 import net.kyori.adventure.text.Component;
-import org.bukkit.permissions.Permissible;
 import org.jetbrains.annotations.NotNull;
-import org.jetbrains.annotations.Nullable;
 import org.jetbrains.annotations.Unmodifiable;
 
 import java.util.Optional;
@@ -48,7 +43,7 @@ public interface PaperPlayerClient extends PlayerClient<PaperPermission, PaperPl
     @NotNull Optional<Suffix> getSuffix(@NotNull String name, @NotNull Context context);
 
     @NotNull
-    Optional<Suffix> setSuffix(@NotNull String name, @NotNull Component content, Context context);
+    Optional<Suffix> setSuffix(@NotNull String name, @NotNull Component content, @NotNull Context context);
 
     boolean deleteSuffix(@NotNull String name, @NotNull Context context);
 
diff --git a/src/main/java/net/kissenpvp/paper/api/nms/NMS.java b/src/main/java/net/kissenpvp/paper/api/nms/NMS.java
index b772f97646bb24146f4aa0a67ae3da852d8bfd27..7fbad0fe6efd8fd024caac234d1a52ee056b76b9 100644
--- a/src/main/java/net/kissenpvp/paper/api/nms/NMS.java
+++ b/src/main/java/net/kissenpvp/paper/api/nms/NMS.java
@@ -18,14 +18,15 @@ package net.kissenpvp.paper.api.nms;
 
 import net.kissenpvp.core.api.base.Implementation;
 import net.kissenpvp.core.api.reflection.ReflectionClass;
+import org.jetbrains.annotations.NotNull;
 
 
 public interface NMS extends Implementation
 {
 
-    String getVersion();
+    @NotNull String getVersion();
 
-    ReflectionClass getNMSClass(String clazz);
+    @NotNull ReflectionClass getNMSClass(@NotNull String clazz);
 
-    ReflectionClass getBukkitClass(String clazz);
+    @NotNull ReflectionClass getBukkitClass(@NotNull String clazz);
 }
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 7674d3354b74823845668e02d4b50bcc87a512cd..b01b428473ddc7fc47638f5af613530bcfa33a18 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -76,7 +76,7 @@ public final class Bukkit {
     private Bukkit() {}
 
     // KissenPaper start
-    public static PaperKissen getKissen() {
+    public static @NotNull PaperKissen getKissen() {
         return kissen;
     }
 
