From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Wed, 24 Apr 2024 21:30:06 +0200
Subject: [PATCH] Adjusted gradle script to integrate into kissen


diff --git a/build.gradle.kts b/build.gradle.kts
index b7d584c8b96eb7699e5df90ed04bf22ee256d571..ab8b7bce37f6b4fc710268ee44f94f303493c837 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -58,6 +58,10 @@ dependencies {
     implementation("org.ow2.asm:asm-commons:9.7")
     // Paper end
 
+    // Pulvinar start
+    api(project(":kissen-api"))
+    // Pulvinar end
+
     api("org.apache.maven:maven-resolver-provider:3.9.6") // Paper - make API dependency for Paper Plugins
     compileOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.18")
     compileOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.18")
@@ -79,8 +83,8 @@ dependencies {
     testImplementation("org.ow2.asm:asm-tree:9.7")
 }
 
-// Paper start
-val generatedApiPath: java.nio.file.Path = rootProject.projectDir.toPath().resolve("paper-api-generator/generated")
+// Paper start // Pulvinar - adjust generator path
+val generatedApiPath: java.nio.file.Path = rootProject.projectDir.toPath().resolve("Pulvinar/paper-api-generator/generated")
 idea {
     module {
         generatedSourceDirs.add(generatedApiPath.toFile())
