From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivo Quiring <falk.quiring@gmail.com>
Date: Sun, 7 Jan 2024 15:22:01 +0100
Subject: [PATCH] Integrated ban system from kissen into paper


diff --git a/src/main/java/net/kissenpvp/pulvinar/api/ban/BanImplementation.java b/src/main/java/net/kissenpvp/pulvinar/api/ban/BanImplementation.java
new file mode 100644
index 0000000000000000000000000000000000000000..253b87c7b446f403cc1c092da925fdea1149d31f
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/api/ban/BanImplementation.java
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2023 KissenPvP
+ *
+ * This program is licensed under the Apache License, Version 2.0.
+ *
+ * This software may be redistributed and/or modified under the terms
+ * of the Apache License as published by the Apache Software Foundation,
+ * either version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the Apache
+ * License, Version 2.0 for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * You should have received a copy of the Apache License, Version 2.0
+ * along with this program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+
+package net.kissenpvp.pulvinar.api.ban;
+
+import net.kissenpvp.core.api.ban.AbstractBanImplementation;
+import net.kissenpvp.core.api.ban.BanType;
+import net.kissenpvp.core.api.database.meta.BackendException;
+import net.kissenpvp.core.api.networking.client.entitiy.ServerEntity;
+import net.kissenpvp.pulvinar.api.base.Context;
+import net.kyori.adventure.text.Component;
+import org.bukkit.ban.ProfileBanList;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.Unmodifiable;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * Interface for Paper-specific ban implementation.
+ * <p>
+ * This interface extends the {@link AbstractBanImplementation} and specializes the type for {@link BanTemplate} and {@link Punishment}.
+ * This interface further defines methods to get punishment details within a Paper context.
+ *
+ * @see AbstractBanImplementation
+ * @see BanTemplate
+ * @see Punishment
+ */
+public interface BanImplementation extends AbstractBanImplementation<BanTemplate, Punishment>, ProfileBanList
+{
+    /**
+     * Returns an unmodifiable set of all punishments applied in the specified context.
+     *
+     * @param context The context from which to retrieve punishments.
+     * @return An unmodifiable set containing all punishments in the context.
+     * @throws BackendException if an error occurs when retrieving the punishment data.
+     */
+    @NotNull @Unmodifiable Set<Punishment> getPunishmentSet(@NotNull Context context) throws BackendException;
+
+    /**
+     * Returns an unmodifiable set of all punishments applied to the player with the given UUID within the specified context.
+     *
+     * @param totalID The UUID of the player for whom to retrieve the punishment data.
+     * @param context The context from which to retrieve punishments.
+     * @return An unmodifiable set containing all punishments for the player in the context.
+     * @throws BackendException if an error occurs when retrieving the punishment data.
+     */
+    @NotNull @Unmodifiable Set<Punishment> getPunishmentSet(@NotNull UUID totalID, @NotNull Context context) throws BackendException;
+
+    /**
+     * Returns the most recent punishment applied to the player with the given UUID within the specified context, if such a punishment exists.
+     *
+     * @param totalID The UUID of the player for whom to retrieve the latest punishment.
+     * @param context The context from which to retrieve the punishment.
+     * @return An Optional containing the latest punishment if it exists, or Optional.empty() otherwise.
+     * @throws BackendException if an error occurs when retrieving the punishment data.
+     */
+    @NotNull Optional<Punishment> getLatestPunishment(@NotNull UUID totalID, @NotNull Context context) throws BackendException;
+
+    /**
+     * Returns the most recent punishment of the specified type applied to the player with the given UUID within the specified context, if such a punishment exists.
+     *
+     * @param totalID The UUID of the player for whom to retrieve the latest punishment.
+     * @param banType The type of ban to search for.
+     * @param context The context from which to retrieve the punishment.
+     * @return An Optional containing the latest punishment of the required type if it exists, or Optional.empty() otherwise.
+     * @throws BackendException if an error occurs when retrieving the punishment data.
+     */
+    @NotNull Optional<Punishment> getLatestPunishment(@NotNull UUID totalID, @NotNull BanType banType, @NotNull Context context) throws BackendException;
+
+    /**
+     * Applies a punishment to the player with the specified UUID.
+     * <p>
+     * This method creates and returns a new instance of {@link Punishment} on the specified player,
+     * based on the given parameters. The new punishment is persistently stored in the backend.
+     *
+     * @param totalID The UUID of the player to be punished.
+     * @param ban The {@link BanTemplate} instance representing the specifications of the punishment.
+     * @param banOperator The operator who is applying the punishment.
+     * @param reason The reason for the punishment. This can be null.
+     * @param context The context in which the punishment is being applied.
+     * @return The created {@link Punishment} instance.
+     * @throws BackendException if an error occurs when trying to create or persist the punishment in the backend.
+     */
+    @NotNull
+    Punishment punish(@NotNull UUID totalID, @NotNull BanTemplate ban, @NotNull ServerEntity banOperator, @Nullable Component reason, @NotNull Context context) throws BackendException;
+
+    @NotNull
+    Punishment punish(@NotNull UUID totalID, @NotNull BanTemplate ban, @NotNull ServerEntity banOperator, @NotNull Context context) throws BackendException;
+
+    @NotNull
+    Punishment punish(@NotNull UUID totalID, @NotNull BanTemplate ban, @NotNull ServerEntity banOperator, boolean apply, @NotNull Context context) throws BackendException;
+
+    @NotNull
+    Punishment punish(@NotNull UUID totalID, @NotNull BanTemplate ban, @NotNull ServerEntity banOperator, boolean apply, @Nullable Component reason, @NotNull Context context) throws BackendException;
+}
diff --git a/src/main/java/net/kissenpvp/pulvinar/api/ban/BanParser.java b/src/main/java/net/kissenpvp/pulvinar/api/ban/BanParser.java
new file mode 100644
index 0000000000000000000000000000000000000000..d362716fe937105e883888fe01dca3483075920e
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/api/ban/BanParser.java
@@ -0,0 +1,60 @@
+package net.kissenpvp.pulvinar.api.ban;
+
+import net.kissenpvp.core.api.ban.AbstractBanTemplate;
+import net.kissenpvp.core.api.ban.exception.NonexistentBanException;
+import net.kissenpvp.core.api.command.CommandPayload;
+import net.kissenpvp.core.api.command.exception.deserialization.DeserializationException;
+import net.kissenpvp.pulvinar.api.command.ArgumentParser;
+import net.kissenpvp.pulvinar.api.command.CommandExceptionHandler;
+import net.kyori.adventure.text.Component;
+import org.bukkit.Bukkit;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.Objects;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class BanParser implements ArgumentParser<BanTemplate>, CommandExceptionHandler<NonexistentBanException> {
+
+    private static @NotNull BanImplementation getBanImplementation() {
+        return Bukkit.getKissen().getImplementation(BanImplementation.class);
+    }
+
+    @Override
+    public @NotNull String serialize(@NotNull BanTemplate object) {
+        return String.valueOf(object.getID());
+    }
+
+    @Override
+    public @NotNull BanTemplate deserialize(@NotNull String input) {
+        BanImplementation banImplementation = getBanImplementation();
+
+        try {
+            return banImplementation.getBanTemplate(Integer.parseInt(input)).orElseThrow(NumberFormatException::new);
+        } catch (NumberFormatException ignored) {
+            Predicate<BanTemplate> equals = (ban) -> Objects.equals(ban.getName(), input);
+            Supplier<DeserializationException> notExisting = () -> new DeserializationException(new NonexistentBanException(input));
+            return banImplementation.getBanTemplates().stream().filter(equals).findFirst().orElseThrow(notExisting);
+        }
+    }
+
+    @Override
+    public @Nullable String argumentName() {
+        return "banid";
+    }
+
+    @Override
+    public @NotNull Collection<String> tabCompletion(@NotNull CommandPayload<CommandSender> commandPayload) {
+        return getBanImplementation().getBanTemplates().stream().map(AbstractBanTemplate::getID).map(String::valueOf).collect(Collectors.toSet());
+    }
+
+    @Override
+    public boolean handle(@NotNull CommandPayload<CommandSender> commandPayload, @NotNull NonexistentBanException throwable) {
+        commandPayload.getSender().sendMessage(Component.translatable("server.ban.invalid", Component.text(throwable.getBan())));
+        return true;
+    }
+}
diff --git a/src/main/java/net/kissenpvp/pulvinar/api/ban/BanTemplate.java b/src/main/java/net/kissenpvp/pulvinar/api/ban/BanTemplate.java
new file mode 100644
index 0000000000000000000000000000000000000000..66ede44eed5358fc35b370202cd48692d4179626
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/api/ban/BanTemplate.java
@@ -0,0 +1,33 @@
+/*
+ * Copyright (C) 2023 KissenPvP
+ *
+ * This program is licensed under the Apache License, Version 2.0.
+ *
+ * This software may be redistributed and/or modified under the terms
+ * of the Apache License as published by the Apache Software Foundation,
+ * either version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the Apache
+ * License, Version 2.0 for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * You should have received a copy of the Apache License, Version 2.0
+ * along with this program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+
+package net.kissenpvp.pulvinar.api.ban;
+
+import net.kissenpvp.core.api.ban.AbstractBanTemplate;
+
+/**
+ * Interface for a Paper-specific ban.
+ * <p>
+ * This interface extends the {@link AbstractBanTemplate} and doesn't introduce any new methods; it merely represents a
+ * specialized version of a ban within the context of the Paper API.
+ * <p>
+ * All methods inherited from {@link AbstractBanTemplate} maintain their original functionality in the Paper API context.
+ *
+ * @see AbstractBanTemplate
+ */
+public interface BanTemplate extends AbstractBanTemplate {}
diff --git a/src/main/java/net/kissenpvp/pulvinar/api/ban/Punishment.java b/src/main/java/net/kissenpvp/pulvinar/api/ban/Punishment.java
new file mode 100644
index 0000000000000000000000000000000000000000..cbd4f5d6edc3cbf8c1b190268098d7163d3fb8a7
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/api/ban/Punishment.java
@@ -0,0 +1,38 @@
+/*
+ * Copyright (C) 2023 KissenPvP
+ *
+ * This program is licensed under the Apache License, Version 2.0.
+ *
+ * This software may be redistributed and/or modified under the terms
+ * of the Apache License as published by the Apache Software Foundation,
+ * either version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the Apache
+ * License, Version 2.0 for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * You should have received a copy of the Apache License, Version 2.0
+ * along with this program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+
+package net.kissenpvp.pulvinar.api.ban;
+
+import com.destroystokyo.paper.profile.PlayerProfile;
+import net.kissenpvp.core.api.ban.AbstractPunishment;
+import org.bukkit.BanEntry;
+import org.bukkit.entity.Player;
+
+/**
+ * Represents a punishment applied to a Player within Paper-specific context.
+ * <p>
+ * This interface is an extension of the {@link AbstractPunishment} interface specific to Player instances.
+ * It defines the contract for any kind of punishment that can be applied to a Player in the game
+ * or application built on the Paper platform.
+ * <p>
+ * The specific methods of punishment and their behaviour on a Player are defined in the classes that
+ * implement this interface, in accordance with the specifications of the {@link AbstractPunishment} interface.
+ *
+ * @see AbstractPunishment
+ */
+public interface Punishment extends AbstractPunishment<Player>, BanEntry<PlayerProfile> {}
diff --git a/src/main/java/net/kissenpvp/pulvinar/api/ban/PunishmentParser.java b/src/main/java/net/kissenpvp/pulvinar/api/ban/PunishmentParser.java
new file mode 100644
index 0000000000000000000000000000000000000000..5d3a48e74b3e17fddb993f9a2eaeaa58700fb243
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/api/ban/PunishmentParser.java
@@ -0,0 +1,44 @@
+package net.kissenpvp.pulvinar.api.ban;
+
+import net.kissenpvp.core.api.ban.AbstractPunishment;
+import net.kissenpvp.core.api.ban.exception.NonexistentPunishmentException;
+import net.kissenpvp.core.api.command.CommandPayload;
+import net.kissenpvp.pulvinar.api.command.ArgumentParser;
+import net.kissenpvp.pulvinar.api.command.CommandExceptionHandler;
+import net.kyori.adventure.text.Component;
+import org.bukkit.Bukkit;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+public class PunishmentParser implements ArgumentParser<Punishment>, CommandExceptionHandler<NonexistentPunishmentException> {
+    @Override
+    public @NotNull String serialize(@NotNull Punishment object) {
+        return object.getID();
+    }
+
+    @Override
+    public @NotNull Punishment deserialize(@NotNull String input) {
+        return Bukkit.getKissen().getImplementation(BanImplementation.class).getPunishmentSet().stream().filter(punishment -> punishment.getID().equals(input)).findFirst().orElseThrow(() -> new NonexistentPunishmentException(input));
+    }
+
+    @Override
+    public @Nullable String argumentName()
+    {
+        return "punishmentid";
+    }
+
+    @Override
+    public @NotNull Collection<String> tabCompletion(@NotNull CommandPayload<CommandSender> commandPayload) {
+        return Bukkit.getKissen().getImplementation(BanImplementation.class).getPunishmentSet().stream().map(AbstractPunishment::getID).collect(Collectors.toSet());
+    }
+
+    @Override
+    public boolean handle(@NotNull CommandPayload<CommandSender> commandPayload, @NotNull NonexistentPunishmentException throwable) {
+        commandPayload.getSender().sendMessage(Component.translatable("server.ban.punishment.invalid", Component.text(throwable.getPunishment())));
+        return true;
+    }
+}
diff --git a/src/main/java/net/kissenpvp/pulvinar/api/ban/warn/Warn.java b/src/main/java/net/kissenpvp/pulvinar/api/ban/warn/Warn.java
new file mode 100644
index 0000000000000000000000000000000000000000..6683bcfc6a28b31621999d80669b2849105606ef
--- /dev/null
+++ b/src/main/java/net/kissenpvp/pulvinar/api/ban/warn/Warn.java
@@ -0,0 +1,6 @@
+package net.kissenpvp.pulvinar.api.ban.warn;
+
+import net.kissenpvp.core.api.user.AbstractWarn;
+import net.kissenpvp.pulvinar.api.ban.BanTemplate;
+
+public interface Warn extends AbstractWarn<BanTemplate> {}
diff --git a/src/main/java/net/kissenpvp/pulvinar/api/networking/client/entity/PulvinarPlayerClient.java b/src/main/java/net/kissenpvp/pulvinar/api/networking/client/entity/PulvinarPlayerClient.java
index cccb508701012e50dd41764950db1e8e1fdcd42f..256248f8a398c0026605b0a1bf18ff987583bd27 100644
--- a/src/main/java/net/kissenpvp/pulvinar/api/networking/client/entity/PulvinarPlayerClient.java
+++ b/src/main/java/net/kissenpvp/pulvinar/api/networking/client/entity/PulvinarPlayerClient.java
@@ -19,17 +19,24 @@
 package net.kissenpvp.pulvinar.api.networking.client.entity;
 
 import net.kissenpvp.core.api.networking.client.entitiy.PlayerClient;
+import net.kissenpvp.core.api.networking.client.entitiy.ServerEntity;
 import net.kissenpvp.core.api.time.AccurateDuration;
 import net.kissenpvp.core.api.user.User;
+import net.kissenpvp.pulvinar.api.ban.BanTemplate;
+import net.kissenpvp.pulvinar.api.ban.Punishment;
+import net.kissenpvp.pulvinar.api.ban.warn.Warn;
 import net.kissenpvp.pulvinar.api.permission.PaperPermissible;
 import net.kissenpvp.pulvinar.api.permission.PermissibleOverriddenException;
 import net.kissenpvp.pulvinar.api.base.Context;
 import net.kissenpvp.pulvinar.api.user.playersetting.BoundPlayerSetting;
 import net.kissenpvp.pulvinar.api.user.playersetting.PlayerSetting;
 import net.kissenpvp.pulvinar.api.user.rank.PlayerRank;
+import net.kyori.adventure.text.Component;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 import org.jetbrains.annotations.Unmodifiable;
 
+import java.util.List;
 import java.util.Set;
 
 public interface PulvinarPlayerClient extends PlayerClient<PlayerRank, Punishment> {
@@ -48,6 +55,13 @@ public interface PulvinarPlayerClient extends PlayerClient<PlayerRank, Punishmen
 
     <X> @NotNull BoundPlayerSetting<X> getSetting(@NotNull Class<? extends PlayerSetting<X>> settingClass, @NotNull Context context);
 
+    @NotNull
+    Warn warn(@NotNull BanTemplate ban, @NotNull ServerEntity warnOperator);
+
+    @NotNull Warn warn(@NotNull BanTemplate ban, @NotNull ServerEntity warnOperator, @Nullable Component reason);
+
+    @NotNull @Unmodifiable
+    List<Warn> getWarnHistory();
 
     void clearActiveWarns();
 }
diff --git a/src/main/java/org/bukkit/event/Event.java b/src/main/java/org/bukkit/event/Event.java
index 8ec56cd6b8e0f5c5dd8c7c88b4671e18dcf109d0..e2db4b4e3c04b408a5ff5d5a464b2b7f26d44a46 100644
--- a/src/main/java/org/bukkit/event/Event.java
+++ b/src/main/java/org/bukkit/event/Event.java
@@ -12,7 +12,8 @@ import org.jetbrains.annotations.NotNull;
  * @see PluginManager#callEvent(Event)
  * @see PluginManager#registerEvents(Listener,Plugin)
  */
-public abstract class Event {
+public abstract class Event implements net.kissenpvp.core.api.event.EventClass
+{
     private String name;
     private final boolean async;
 
